Blog: Niklas on fuzzing (#111)
bitcoinbrink/website: _posts/2023-07-05-niklas-fuzzing.md

---
title: "Niklas Gögge on Fuzz Testing"
permalink: /blog/2023/07/05/fuzzing/
layout: post
author: brink
name: "brink"
alt: Niklas Gögge on Fuzz Testing
category: "News"
description: Niklas Gögge on Fuzz Testing.
---

In a security-focused project like Bitcoin Core, testing is critical. [Bitcoin
Core unit tests][] are written in C++ using the Boost libraries to quickly test
components of the codebase in isolation (see the [bech32 unit tests][] for
example). [Bitcoin Core functional tests][] are written in python to test
features of the software, involving multiple components working together (see
the [P2P ping functional tests][]).

> A less-familiar type of testing is fuzz testing where software repeatedly calls

Reviewer: Why "less-familiar" and not "more cutting-edge" or something?

I think this description can be a lot more compelling. Imo calling patterns are not as important as the fact that (1)Fuzzing is characterized by the fact that inputs are (hopefully intelligently) generated by the fuzzing engine (2) fuzzing scales wayyy better than human code review, you can often just throw more money at it to find more bugs.

into portions of the Bitcoin Core codebase with random, potentially
malformed or semi-valid inputs in order to surface potential negative consequences

> such as a crash, memory leak, or other failure scenarios.

Reviewer: Could provide some examples of theoretical bugs that can be found, compelling but easy enough to understand.

"For example, fuzzing can uncover a series of specially-crafted p2p messages that leads to a crash, memory leak, or inconsistent internal state. As Bitcoin Core nodes are constantly exposed to anonymous peers on the network, a crash bug in the p2p message processing code can be catastrophic."

Reviewer 2: Could also link to https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Fuzz-Trophies, which is outdated now but still has a good list of old issues that were found.

Author: That’s great, added, thanks @dergoegge !

---

In a security-focused project like Bitcoin Core, testing is critical. [Bitcoin
Core unit tests][] are written in C++ using the Boost libraries to quickly test
components of the codebase in isolation (see the [bech32 unit tests][] for
example). [Bitcoin Core functional tests][] are written in python to test
features of the software, involving multiple components working together (see
the [P2P ping functional tests][]).

A more cutting-edge type of testing is fuzz testing where software repeatedly calls
into portions of the Bitcoin Core codebase with random, potentially
malformed or semi-valid inputs in order to surface potential negative consequences
such as a crash, memory leak, or other failure scenarios. When done
continuously, these auto-generated mutated inputs from the fuzzing engine can be
used to expand the test coverage and uncover bugs.

For example, fuzzing can uncover a series of specially-crafted P2P messages that
leads to a crash, memory leak, or inconsistent internal state. As Bitcoin Core
nodes are constantly exposed to anonymous peers on the network, a crash bug in
the P2P message processing code could be catastrophic.

> In contrast to manual testing or individual code review, fuzz testing scales
> with the amount of computation allocated to it and can be running 24/7. Bitcoin
> Core participates in [OSS-Fuzz][], Google's continuous fuzzing program, in
> addition to contributors running fuzzing infrastructure and contributing new
> inputs back to the growing body of [fuzzing corpora][].

Reviewer 2: ```suggestion
As a form of automated software testing, fuzzing can scale (to some extend) with the amount of computation allocated to it and if done continuously often finds edge cases that were missed during manual code review. Bitcoin
Core participates in [OSS-Fuzz][], Google's continuous fuzzing program, in
addition to contributors running fuzzing infrastructure and contributing new
inputs back to the growing body of [fuzzing corpora][].
```

Author: Thanks @dergoegge , took this suggestion with a small change

---
title: "Niklas Gögge on Fuzz Testing"
permalink: /blog/2023/07/05/fuzzing/
layout: post
author: brink
name: "brink"
alt: Niklas Gögge on Fuzz Testing

> category: "News"

Reviewer: Is this the right category?

Author: I dont think that category is currently used, but yes, this would be in the technical category, Ill update

description: Niklas Gögge on Fuzz Testing.
---

In a security-focused project like Bitcoin Core, testing is critical. [Bitcoin
Core unit tests][] are written in C++ using the Boost libraries to quickly test
components of the codebase in isolation (see the [bech32 unit tests][] for
example). [Bitcoin Core functional tests][] are written in python to test
features of the software, involving multiple components working together (see
the [P2P ping functional tests][]).

A more cutting-edge type of testing is fuzz testing where software repeatedly calls

> into portions of the Bitcoin Core codebase with random, potentially
> malformed or semi-valid inputs in order to surface potential negative consequences
> such as a crash, memory leak, or other failure scenarios. When done
> continuously, these auto-generated mutated inputs from the fuzzing engine can be
> used to expand the test coverage and uncover bugs.
> 

Reviewer: I think we can deduplicate a bit and combine this with the subsequent paragraph.
```suggestion
into portions of the codebase using inputs generated automatically by a fuzzing engine
instead of manually by the test writer. When done
continuously, fuzzing can be
used to expand test coverage and uncover bugs.
```

> For example, fuzzing can uncover a series of specially-crafted P2P messages that

Reviewer: To not use "uncover" twice in quick succession
```suggestion
For example, fuzzing can generate a series of specially-crafted P2P messages that
```

leads to a crash, memory leak, or inconsistent internal state. As Bitcoin Core
nodes are constantly exposed to anonymous peers on the network, a crash bug in
the P2P message processing code could be catastrophic.

As a form of automated software testing, fuzzing can scale (to some extent) with

> the amount of computation allocated to it and if done continuously often finds

Reviewer: ```suggestion
the amount of computation allocated to it and often finds
```

edge cases that were missed during manual code review. Bitcoin
Core participates in [OSS-Fuzz][], Google's continuous fuzzing program, in
addition to contributors running fuzzing infrastructure and contributing new
inputs back to the growing body of [fuzzing corpora][].

Fuzz testing initiatives in Bitcoin Core have found several important bugs over the years,

> with [Fuzz Trophies][] noting some of them.

Reviewer: ```suggestion
with [Fuzz Trophies][] documenting some of the less severe ones.
```

Brink grantee [Niklas Gögge][niklas github] has been working on [Bitcoin Core's
fuzz testing][] initiatives including increasing the number of fuzz-testable

> [targets][fuzz targets] (~200 currently) and refactoring existing areas of the
> codebase, like [P2P net processing][fuzzing net processing], to be more fuzzer-friendly.

Reviewer: A bit less verbose and AFAIK Niklas is only refactoring p2p
```suggestion
[targets][fuzz targets] (~200 currently) and refactoring 
[P2P net processing][fuzzing net processing] to be more fuzzer-friendly.
```
